:NoJS
def foo: ~int\x
//│ foo: ~int\x

foo: ~(int\x)
//│ res: ~int\x

foo: (~int)\x
//│ res: ~int

res: ~(int\x)
//│ res: ~int\x


def foo: (~ 'a \ x) -> 'a -> 'a
//│ foo: ~'a\x -> 'a -> 'a

foo 1
//│ res: ('a & ~1) -> 'a

def foo: (~ 'a \ x) -> 'a
//│ foo: anything -> nothing

foo 1
//│ res: nothing


class X: {}
//│ Defined class X

def v0: (~(X\x))\y
//│ v0: ~X\x\y

:e
v0: int
//│ ╔══[ERROR] Type mismatch in type ascription:
//│ ║  l.35: 	v0: int
//│ ║        	^^
//│ ╟── type `~X\x\y` does not match type `int`
//│ ║  l.31: 	def v0: (~(X\x))\y
//│ ║        	        ^^^^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.35: 	v0: int
//│ ║        	^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.35: 	v0: int
//│ ╙──      	    ^^^
//│ res: int



// def test: (int | { x: 'a } | { y: 'a }) -> 'a
def test: (int & { f: 'a } | { x: 'a } | { y: 'a }) -> 'a
//│ test: anything -> nothing

:e
def test x = case x of { int -> x.f | _ -> x.x }
//│ (int & {f: 'a} | {x: 'a} & ~int) -> 'a
//│   <:  test:
//│ anything -> nothing
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ║        	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── expression of type `int & {x: 'a} | {x: 'a} & ~?a` does not have field 'f'
//│ ╟── Note: constraint arises from field selection:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ║        	                                ^^^
//│ ╟── from refined scrutinee:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ╙──      	                  ^
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ║        	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── expression of type `{y: 'a} & ~int` does not have field 'x'
//│ ╟── Note: constraint arises from field selection:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ╙──      	                                           ^^^
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ║        	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── expression of type `int & {y: 'a} | {y: 'a} & ~?a` does not have field 'f'
//│ ╟── Note: constraint arises from field selection:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ║        	                                ^^^
//│ ╟── from refined scrutinee:
//│ ║  l.57: 	def test x = case x of { int -> x.f | _ -> x.x }
//│ ╙──      	                  ^

// Note: not an error, but a terminating `test` can't be implemented
test 1
//│ res: nothing

test { x = 1 }
//│ res: nothing

test { y = 1 }
//│ res: nothing


class A: { }
class B: { }
class C: { }
//│ Defined class A
//│ Defined class B
//│ Defined class C

def test: (A & { a: 'a } | B & { b: 'a } | C & { c: 'a }) -> 'a
//│ test: (A & {a: 'a} | B & {b: 'a} | C & {c: 'a}) -> 'a

test (error: A & { a: int })
test (error: B & { b: int })
test (error: C & { c: int })
//│ res: int
//│ res: int
//│ res: int

:e
test (error: B & { c: int })
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.118: 	test (error: B & { c: int })
//│ ║         	^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── type `B & {c: int}` does not match type `A & {a: ?a} | B & {b: ?a} | C & {c: ?a}`
//│ ║  l.118: 	test (error: B & { c: int })
//│ ║         	             ^^^^^^^^^^^^^^
//│ ╟── Note: constraint arises from union type:
//│ ║  l.107: 	def test: (A & { a: 'a } | B & { b: 'a } | C & { c: 'a }) -> 'a
//│ ╙──       	          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ res: error


def test: (A & { x: 'a; y: 'b } | B & { x: 'a; y: 'b } | C & { x: 'a; y: 'b }) -> { l: 'a; r: 'b }
//│ test: (A & {x: 'a, y: 'b} | B & {x: 'a, y: 'b} | C & {x: 'a, y: 'b}) -> {l: 'a, r: 'b}

test (error: A & { x: int; y: string })
test (error: B & { x: int; y: string })
test (error: C & { x: int; y: string })
//│ res: {l: int, r: string}
//│ res: {l: int, r: string}
//│ res: {l: int, r: string}

:e
test (error: A & { x: int })
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.142: 	test (error: A & { x: int })
//│ ║         	^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── type `A & {x: int}` does not match type `A & {x: ?a, y: ?b} | B & {x: ?a, y: ?b} | C & {x: ?a, y: ?b}`
//│ ║  l.142: 	test (error: A & { x: int })
//│ ║         	             ^^^^^^^^^^^^^^
//│ ╟── Note: constraint arises from union type:
//│ ║  l.131: 	def test: (A & { x: 'a; y: 'b } | B & { x: 'a; y: 'b } | C & { x: 'a; y: 'b }) -> { l: 'a; r: 'b }
//│ ╙──       	          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ res: error | {l: int, r: nothing}


def test: (A & { xA: 'a; yA: 'b } | B & { xB: 'a; yB: 'b } | C & { xC: 'a; yC: 'b }) -> { l: 'a; r: 'b }
//│ test: (B & {xB: 'a, yB: 'b} | A & {xA: 'a, yA: 'b} | C & {xC: 'a, yC: 'b}) -> {l: 'a, r: 'b}

test (error: A & { xA: int; yA: string })
test (error: B & { xB: int; yB: string })
test (error: C & { xC: int; yC: string })
//│ res: {l: int, r: string}
//│ res: {l: int, r: string}
//│ res: {l: int, r: string}


def negInt: ~int
//│ negInt: ~int

def v = negInt with { x = 1 }
//│ v: {x: 1} & ~int

v.x
//│ res: 1

(id v).x
//│ res: 1

:e
negInt: A | { x: 'a }
//│ ╔══[ERROR] Type mismatch in type ascription:
//│ ║  l.179: 	negInt: A | { x: 'a }
//│ ║         	^^^^^^
//│ ╟── type `~int` does not match type `{x: ?a} | A`
//│ ║  l.166: 	def negInt: ~int
//│ ║         	            ^^^^
//│ ╟── but it flows into reference with expected type `{x: ?a} | A`
//│ ║  l.179: 	negInt: A | { x: 'a }
//│ ║         	^^^^^^
//│ ╟── Note: constraint arises from union type:
//│ ║  l.179: 	negInt: A | { x: 'a }
//│ ╙──       	        ^^^^^^^^^^^^^
//│ res: {x: nothing} | A

v: A | { x: 'a }
//│ res: {x: 1} | A

(error: A) with { x = 1 }
//│ res: A\x & {x: 1}

// Note that we get {x: nothing} because the LHS is a subtype of A so 'a is not constrained
(error: A) with { x = 1 } : A | { x: 'a }
//│ res: {x: nothing} | A

id (error: A) with { x = 1 } : A | { x: 'a }
//│ res: {x: nothing} | A

def negWeird: ~(~(~(A & { x: int })))
//│ negWeird: ~A

def v = negWeird with { x = 1 }
//│ v: {x: 1} & ~a

v: A | { x: 'a }
//│ res: {x: 1} | A


def funny = error: nothing\x
//│ funny: nothing

// It could be agrued this should fail... but that would make constraint solving more awkward!
funny.x
//│ res: nothing


:e
(error: { x: int }): int | string
//│ ╔══[ERROR] Type mismatch in type ascription:
//│ ║  l.226: 	(error: { x: int }): int | string
//│ ║         	^^^^^^^^^^^^^^^^^^^
//│ ╟── type `{x: int}` does not match type `int | string`
//│ ║  l.226: 	(error: { x: int }): int | string
//│ ║         	        ^^^^^^^^^^
//│ ╟── Note: constraint arises from union type:
//│ ║  l.226: 	(error: { x: int }): int | string
//│ ╙──       	                     ^^^^^^^^^^^^
//│ res: int | string


(error: (~ 'a) \ x): { x: int }
//│ res: {x: int}


idx = id with { x = 1 }
//│ idx: 'a -> 'a & {x: 1}

idx idx.x
//│ res: 1

def indirect i = i i.x
//│ indirect: ('a -> 'b & {x: 'a}) -> 'b

indirect idx
//│ res: 1

if true then id else idx
//│ res: 'a -> 'a

r =  if true then id with { x = "a"; y = "b" } else idx
//│ r: 'a -> 'a & {x: "a", y: "b"}

r r.x
//│ res: "a" | 1

indirect r
//│ res: "a" | 1

def takeFunOrRecord: (int -> int | { x: int }) -> int
//│ takeFunOrRecord: anything -> int

:pe
takeFunOrRecord x = case x of { { x = v } -> 0 | _ -> 1 }
//│ /!\ Parse error: Expected "}":1:33, found "{ x = v } " at l.271:33: takeFunOrRecord x = case x of { { x = v } -> 0 | _ -> 1 }

:pe
takeFunOrRecord x = case x of { { x } -> 0 | _ -> 1 }
//│ /!\ Parse error: Expected "}":1:33, found "{ x } -> 0" at l.275:33: takeFunOrRecord x = case x of { { x } -> 0 | _ -> 1 }

:pe
takeFunOrRecord x = case x of { (int -> int) -> 0 | _ -> 1 }
//│ /!\ Parse error: Expected "}":1:33, found "(int -> in" at l.279:33: takeFunOrRecord x = case x of { (int -> int) -> 0 | _ -> 1 }

:pe
takeFunOrRecord x = case x of { (->) -> 0 | _ -> 1 }
//│ /!\ Parse error: Expected "}":1:33, found "(->) -> 0 " at l.283:33: takeFunOrRecord x = case x of { (->) -> 0 | _ -> 1 }


class AA
trait TT
def ty: AA | ('a \ x)
def ty2: TT & { x: int }
// def ty: AA | (int -> int)
//│ Defined class AA
//│ Defined trait TT
//│ ty: AA
//│ ty2: TT & {x: int}

:e
ty = ty2
//│ TT & {x: int}
//│   <:  ty:
//│ AA
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.298: 	ty = ty2
//│ ║         	^^^^^^^^
//│ ╟── type `TT & {x: int}` does not match type `'a\x | AA`
//│ ║  l.290: 	def ty2: TT & { x: int }
//│ ║         	         ^^^^^^^^^^^^^^^
//│ ╟── but it flows into reference with expected type `'a\x | AA`
//│ ║  l.298: 	ty = ty2
//│ ║         	     ^^^
//│ ╟── Note: constraint arises from union type:
//│ ║  l.289: 	def ty: AA | ('a \ x)
//│ ╙──       	        ^^^^^^^^^^^^^


